paimon-core/src/main/java/org/apache/paimon/table/source/DataTableBatchScan.java

```diff
@@ -96,6 +96,27 @@ private StartingScanner.Result applyPushDownLimit(StartingScanner.Result result)
             long scannedRowCount = 0;
             SnapshotReader.Plan plan = ((ScannedResult) result).plan();
             List<DataSplit> splits = plan.dataSplits();
+            if (splits.isEmpty()) {
+                return result;
+            }
+
+            // We first add the rawConvertible split to avoid merging, and if the row count
+            // is still less than limit number, then add split which is not rawConvertible.
+            splits.sort(
+                    (x, y) -> {
+                        if (x.rawConvertible() && y.rawConvertible()) {
+                            return 0;
+                        } else if (x.rawConvertible()) {
+                            return -1;
+                        } else {
+                            return 1;
+                        }
+                    });
+            // fast return if there is no rawConvertible split
```

Comment:
Why you can ignore pushDownLimit if there is no rawConvertible split?

Reply:
It follows previous behavior, if all splits are not rawConvertible, then we add all splits as the final plan. The semantics of pushDownLimit is just a best effort, the engine will do limit again.

Reply:
OK.